dvc/remote/gdrive/client.py

```diff
@@ -0,0 +1,284 @@
+
+
+class GDriveClient:
+
+    GOOGLEAPIS_BASE_URL = "https://www.googleapis.com/"
+    TIMEOUT = (5, 60)
+
+    def __init__(
+        self,
+        space,
+        oauth_id,
+        credentialpath,
+        scopes,
+        oauth2_flow_runner,
+        max_retries=10,
+    ):
+        self.space = space
+        self.oauth_id = oauth_id
+        self.credentialpath = credentialpath
+        self.scopes = scopes
+        self.oauth2_flow_runner = oauth2_flow_runner
+        self.max_retries = max_retries
+        self.oauth2 = OAuth2(
+            oauth_id, credentialpath, scopes, oauth2_flow_runner
+        )
+
+    @cached_property
+    def session(self):
+        """AuthorizedSession to communicate with https://googleapis.com
+
+        Security notice:
+
+        It always adds the Authorization header to the requests, not paying
+        attention is request is for googleapis.com or not. It is just how
+        AuthorizedSession from google-auth implements adding its headers. Don't
+        use RemoteGDrive.session() to send requests to domains other than
+        googleapis.com.
+        """
+        return self.oauth2.get_session()
+
+    def request(self, method, path, *args, **kwargs):
+        # Google Drive has tight rate limits, which strikes the
+        # performance and gives the 403 and 429 errors.
+        # See https://developers.google.com/drive/api/v3/handle-errors
+        retries = 0
+        exponential_backoff = 1
+        if "timeout" not in kwargs:
+            kwargs["timeout"] = self.TIMEOUT
+        while retries < self.max_retries:
+            retries += 1
+            response = self.session.request(
+                method, self.GOOGLEAPIS_BASE_URL + path, *args, **kwargs
+            )
+            if response_is_ratelimit(response) or response.status_code >= 500:
+                logger.debug(
+                    "got {} response, will retry in {} sec".format(
+                        response.status_code, exponential_backoff
+                    )
+                )
+                sleep(exponential_backoff)
+                exponential_backoff *= 2
+            else:
+                break
+        if response.status_code >= 400:
+            raise GDriveHTTPError(response)
+        return response
+
+    def search(self, parent=None, name=None, add_params={}):
+        query = []
+        if parent is not None:
+            query.append("'{}' in parents".format(parent))
+        if name is not None:
+            query.append("name = '{}'".format(name))
+        params = {"q": " and ".join(query), "spaces": self.space}
+        params.update(add_params)
+        while True:
+            data = self.request("GET", "drive/v3/files", params=params).json()
+            for i in data["files"]:
+                yield i
+            if not data.get("nextPageToken"):
+                break
+            params["pageToken"] = data["nextPageToken"]
+
+    def get_metadata(self, path_info, fields=None):
```

Comment:
could you please add comment what does it return? really hard to understand this. just trying to understand why logic is so complicated here